# Step 1: identifying relevant features and feature values across different files

In [ ]:
# ## Diagnostic codes for ischemic stroke
# ### ICD-9 codes
#
# not ok below
# ----- 433**
# 43300	Occlusion and stenosis of basilar artery without mention of cerebral infarction
# 43301	Occlusion and stenosis of basilar artery with cerebral infarction
# 43310	Occlusion and stenosis of carotid artery without mention of cerebral infarction
# 43311	Occlusion and stenosis of carotid artery with cerebral infarction
# 43320	Occlusion and stenosis of vertebral artery without mention of cerebral infarction
# 43321	Occlusion and stenosis of vertebral artery with cerebral infarction
# 43330	Occlusion and stenosis of multiple and bilateral precerebral arteries without mention of cerebral infarction
# 43331	Occlusion and stenosis of multiple and bilateral precerebral arteries with cerebral infarction
# 43380	Occlusion and stenosis of other specified precerebral artery without mention of cerebral infarction
# 43381	Occlusion and stenosis of other specified precerebral artery with cerebral infarction
# 43390	Occlusion and stenosis of unspecified precerebral artery without mention of cerebral infarction
# 43391	Occlusion and stenosis of unspecified precerebral artery with cerebral infarction
# not ok above
#
# unsure below => with mention of cerebral infarction is ok, without is unsure
# ----- 434**
# 43400	Cerebral thrombosis without mention of cerebral infarction
# 43401	Cerebral thrombosis with cerebral infarction
# 43410	Cerebral embolism without mention of cerebral infarction
# 43411	Cerebral embolism with cerebral infarction
# unsure above
#
# not ok below
# 43490	Cerebral artery occlusion, unspecified without mention of cerebral infarction
# 43491	Cerebral artery occlusion, unspecified with cerebral infarction
# not ok above
#
# unsure below: Supersets of ischemic stroke
# ----- Other
# 436	    Acute, but ill-defined, cerebrovascular disease
# 4370	Cerebral atherosclerosis
# 4371	Other generalized ischemic cerebrovascular disease
# unsure above
#
# 4376    Nonpyogenic thrombosis of intracranial venous sinus - ok (blood clot formation in barin vessels)
#
# ### ICD-10 codes
# ----- I63***
# I63	    10	Cerebral infarction                                               - unsure
# ok below
# I630	10	Cerebral infarction due to thrombosis of precerebral arteries
# I6300	10	Cerebral infarction due to thrombosis of unspecified precerebral artery
# I6301	10	Cerebral infarction due to thrombosis of vertebral artery
# I63011	10	Cerebral infarction due to thrombosis of right vertebral artery
# I63012	10	Cerebral infarction due to thrombosis of left vertebral artery
# I63013	10	Cerebral infarction due to thrombosis of bilateral vertebral arteries
# I63019	10	Cerebral infarction due to thrombosis of unspecified vertebral artery
# I6302	10	Cerebral infarction due to thrombosis of basilar artery
# I6303	10	Cerebral infarction due to thrombosis of carotid artery
# I63031	10	Cerebral infarction due to thrombosis of right carotid artery
# I63032	10	Cerebral infarction due to thrombosis of left carotid artery
# I63033	10	Cerebral infarction due to thrombosis of bilateral carotid arteries
# I63039	10	Cerebral infarction due to thrombosis of unspecified carotid artery
# I6309	10	Cerebral infarction due to thrombosis of other precerebral artery
# I631	10	Cerebral infarction due to embolism of precerebral arteries
# I6310	10	Cerebral infarction due to embolism of unspecified precerebral artery
# I6311	10	Cerebral infarction due to embolism of vertebral artery
# I63111	10	Cerebral infarction due to embolism of right vertebral artery
# I63112	10	Cerebral infarction due to embolism of left vertebral artery
# I63113	10	Cerebral infarction due to embolism of bilateral vertebral arteries
# I63119	10	Cerebral infarction due to embolism of unspecified vertebral artery
# I6312	10	Cerebral infarction due to embolism of basilar artery
# I6313	10	Cerebral infarction due to embolism of carotid artery
# I63131	10	Cerebral infarction due to embolism of right carotid artery
# I63132	10	Cerebral infarction due to embolism of left carotid artery
# I63133	10	Cerebral infarction due to embolism of bilateral carotid arteries
# I63139	10	Cerebral infarction due to embolism of unspecified carotid artery
# I6319	10	Cerebral infarction due to embolism of other precerebral artery
# ok above
#
# not ok below
# I632	10	Cerebral infarction due to unspecified occlusion or stenosis of precerebral arteries
# I6320	10	Cerebral infarction due to unspecified occlusion or stenosis of unspecified precerebral arteries
# I6321	10	Cerebral infarction due to unspecified occlusion or stenosis of vertebral arteries
# I63211	10	Cerebral infarction due to unspecified occlusion or stenosis of right vertebral artery
# I63212	10	Cerebral infarction due to unspecified occlusion or stenosis of left vertebral artery
# I63213	10	Cerebral infarction due to unspecified occlusion or stenosis of bilateral vertebral arteries
# I63219	10	Cerebral infarction due to unspecified occlusion or stenosis of unspecified vertebral artery
# I6322	10	Cerebral infarction due to unspecified occlusion or stenosis of basilar artery
# I6323	10	Cerebral infarction due to unspecified occlusion or stenosis of carotid arteries
# I63231	10	Cerebral infarction due to unspecified occlusion or stenosis of right carotid arteries
# I63232	10	Cerebral infarction due to unspecified occlusion or stenosis of left carotid arteries
# I63233	10	Cerebral infarction due to unspecified occlusion or stenosis of bilateral carotid arteries
# I63239	10	Cerebral infarction due to unspecified occlusion or stenosis of unspecified carotid artery
# I6329	10	Cerebral infarction due to unspecified occlusion or stenosis of other precerebral arteries
# not ok above
#
# ok below
# I633	10	Cerebral infarction due to thrombosis of cerebral arteries
# I6330	10	Cerebral infarction due to thrombosis of unspecified cerebral artery
# I6331	10	Cerebral infarction due to thrombosis of middle cerebral artery
# I63311	10	Cerebral infarction due to thrombosis of right middle cerebral artery
# I63312	10	Cerebral infarction due to thrombosis of left middle cerebral artery
# I63313	10	Cerebral infarction due to thrombosis of bilateral middle cerebral arteries
# I63319	10	Cerebral infarction due to thrombosis of unspecified middle cerebral artery
# I6332	10	Cerebral infarction due to thrombosis of anterior cerebral artery
# I63321	10	Cerebral infarction due to thrombosis of right anterior cerebral artery
# I63322	10	Cerebral infarction due to thrombosis of left anterior cerebral artery
# I63323	10	Cerebral infarction due to thrombosis of bilateral anterior cerebral arteries
# I63329	10	Cerebral infarction due to thrombosis of unspecified anterior cerebral artery
# I6333	10	Cerebral infarction due to thrombosis of posterior cerebral artery
# I63331	10	Cerebral infarction due to thrombosis of right posterior cerebral artery
# I63332	10	Cerebral infarction due to thrombosis of left posterior cerebral artery
# I63333	10	Cerebral infarction due to thrombosis of bilateral posterior cerebral arteries
# I63339	10	Cerebral infarction due to thrombosis of unspecified posterior cerebral artery
# I6334	10	Cerebral infarction due to thrombosis of cerebellar artery
# I63341	10	Cerebral infarction due to thrombosis of right cerebellar artery
# I63342	10	Cerebral infarction due to thrombosis of left cerebellar artery
# I63343	10	Cerebral infarction due to thrombosis of bilateral cerebellar arteries
# I63349	10	Cerebral infarction due to thrombosis of unspecified cerebellar artery
# I6339	10	Cerebral infarction due to thrombosis of other cerebral artery
# I634	10	Cerebral infarction due to embolism of cerebral arteries
# I6340	10	Cerebral infarction due to embolism of unspecified cerebral artery
# I6341	10	Cerebral infarction due to embolism of middle cerebral artery
# I63411	10	Cerebral infarction due to embolism of right middle cerebral artery
# I63412	10	Cerebral infarction due to embolism of left middle cerebral artery
# I63413	10	Cerebral infarction due to embolism of bilateral middle cerebral arteries
# I63419	10	Cerebral infarction due to embolism of unspecified middle cerebral artery
# I6342	10	Cerebral infarction due to embolism of anterior cerebral artery
# I63421	10	Cerebral infarction due to embolism of right anterior cerebral artery
# I63422	10	Cerebral infarction due to embolism of left anterior cerebral artery
# I63423	10	Cerebral infarction due to embolism of bilateral anterior cerebral arteries
# I63429	10	Cerebral infarction due to embolism of unspecified anterior cerebral artery
# I6343	10	Cerebral infarction due to embolism of posterior cerebral artery
# I63431	10	Cerebral infarction due to embolism of right posterior cerebral artery
# I63432	10	Cerebral infarction due to embolism of left posterior cerebral artery
# I63433	10	Cerebral infarction due to embolism of bilateral posterior cerebral arteries
# I63439	10	Cerebral infarction due to embolism of unspecified posterior cerebral artery
# I6344	10	Cerebral infarction due to embolism of cerebellar artery
# I63441	10	Cerebral infarction due to embolism of right cerebellar artery
# I63442	10	Cerebral infarction due to embolism of left cerebellar artery
# I63443	10	Cerebral infarction due to embolism of bilateral cerebellar arteries
# I63449	10	Cerebral infarction due to embolism of unspecified cerebellar artery
# I6349	10	Cerebral infarction due to embolism of other cerebral artery
# ok above
#
# not ok below
# I635	10	Cerebral infarction due to unspecified occlusion or stenosis of cerebral arteries
# I6350	10	Cerebral infarction due to unspecified occlusion or stenosis of unspecified cerebral artery
# I6351	10	Cerebral infarction due to unspecified occlusion or stenosis of middle cerebral artery
# I63511	10	Cerebral infarction due to unspecified occlusion or stenosis of right middle cerebral artery
# I63512	10	Cerebral infarction due to unspecified occlusion or stenosis of left middle cerebral artery
# I63513	10	Cerebral infarction due to unspecified occlusion or stenosis of bilateral middle cerebral arteries
# I63519	10	Cerebral infarction due to unspecified occlusion or stenosis of unspecified middle cerebral artery
# I6352	10	Cerebral infarction due to unspecified occlusion or stenosis of anterior cerebral artery
# I63521	10	Cerebral infarction due to unspecified occlusion or stenosis of right anterior cerebral artery
# I63522	10	Cerebral infarction due to unspecified occlusion or stenosis of left anterior cerebral artery
# I63523	10	Cerebral infarction due to unspecified occlusion or stenosis of bilateral anterior cerebral arteries
# I63529	10	Cerebral infarction due to unspecified occlusion or stenosis of unspecified anterior cerebral artery
# I6353	10	Cerebral infarction due to unspecified occlusion or stenosis of posterior cerebral artery
# I63531	10	Cerebral infarction due to unspecified occlusion or stenosis of right posterior cerebral artery
# I63532	10	Cerebral infarction due to unspecified occlusion or stenosis of left posterior cerebral artery
# I63533	10	Cerebral infarction due to unspecified occlusion or stenosis of bilateral posterior cerebral arteries
# I63539	10	Cerebral infarction due to unspecified occlusion or stenosis of unspecified posterior cerebral artery
# I6354	10	Cerebral infarction due to unspecified occlusion or stenosis of cerebellar artery
# I63541	10	Cerebral infarction due to unspecified occlusion or stenosis of right cerebellar artery
# I63542	10	Cerebral infarction due to unspecified occlusion or stenosis of left cerebellar artery
# I63543	10	Cerebral infarction due to unspecified occlusion or stenosis of bilateral cerebellar arteries
# I63549	10	Cerebral infarction due to unspecified occlusion or stenosis of unspecified cerebellar artery
# I6359	10	Cerebral infarction due to unspecified occlusion or stenosis of other cerebral artery
# not ok above
#
# I636	10	Cerebral infarction due to cerebral venous thrombosis, nonpyogenic      - ok
# I638	10	Other cerebral infarction                                               - unsure
# I6381	10	Other cerebral infarction due to occlusion or stenosis of small artery  - not ok
# I6389	10	Other cerebral infarction                                               - unsure
# I639	10	Cerebral infarction, unspecified                                        - unsure
#
# not ok below
# ----- I65**
# I65	10	Occlusion and stenosis of precerebral arteries, not resulting in cerebral infarction
# I650	10	Occlusion and stenosis of vertebral artery
# I6501	10	Occlusion and stenosis of right vertebral artery
# I6502	10	Occlusion and stenosis of left vertebral artery
# I6503	10	Occlusion and stenosis of bilateral vertebral arteries
# I6509	10	Occlusion and stenosis of unspecified vertebral artery
# I651	10	Occlusion and stenosis of basilar artery
# I652	10	Occlusion and stenosis of carotid artery
# I6521	10	Occlusion and stenosis of right carotid artery
# I6522	10	Occlusion and stenosis of left carotid artery
# I6523	10	Occlusion and stenosis of bilateral carotid arteries
# I6529	10	Occlusion and stenosis of unspecified carotid artery
# I658	10	Occlusion and stenosis of other precerebral arteries
# I659	10	Occlusion and stenosis of unspecified precerebral artery
#
# ----- I66**
# I66	    10	Occlusion and stenosis of cerebral arteries, not resulting in cerebral infarction
# I660	10	Occlusion and stenosis of middle cerebral artery
# I6601	10	Occlusion and stenosis of right middle cerebral artery
# I6602	10	Occlusion and stenosis of left middle cerebral artery
# I6603	10	Occlusion and stenosis of bilateral middle cerebral arteries
# I6609	10	Occlusion and stenosis of unspecified middle cerebral artery
# I661	10	Occlusion and stenosis of anterior cerebral artery
# I6611	10	Occlusion and stenosis of right anterior cerebral artery
# I6612	10	Occlusion and stenosis of left anterior cerebral artery
# I6613	10	Occlusion and stenosis of bilateral anterior cerebral arteries
# I6619	10	Occlusion and stenosis of unspecified anterior cerebral artery
# I662	10	Occlusion and stenosis of posterior cerebral artery
# I6621	10	Occlusion and stenosis of right posterior cerebral artery
# I6622	10	Occlusion and stenosis of left posterior cerebral artery
# I6623	10	Occlusion and stenosis of bilateral posterior cerebral arteries
# I6629	10	Occlusion and stenosis of unspecified posterior cerebral artery
# I663	10	Occlusion and stenosis of cerebellar arteries
# I668	10	Occlusion and stenosis of other cerebral arteries
# I669	10	Occlusion and stenosis of unspecified cerebral artery
# not ok above

### All relevant diagnostic codes
"43300", "43301", "43310", "43311", "43320", "43321", "43330", "43331", "43380", "43381", "43390", "43391", "43400", "43401", "43410", "43411", "43490", "43491", "436", "4370", "4371", "4376", "I63", "I6300", "I6301", "I63011", "I63012", "I63013", "I63019", "I6302", "I6303", "I63031", "I63032", "I63033", "I63039", "I6309", "I631", "I6310", "I6311", "I63111", "I63112", "I63113", "I63119", "I6312", "I6313", "I63131", "I63132", "I63133", "I63139", "I6319", "I632", "I6320", "I6321", "I63211", "I63212", "I63213", "I63219", "I6322", "I6323", "I63231", "I63232", "I63233", "I63239", "I6329", "I633", "I6330", "I6331", "I63311", "I63312", "I63313", "I63319", "I6332", "I63321", "I63322", "I63323", "I63329", "I6333", "I63331", "I63332", "I63333", "I63339", "I6334", "I63341", "I63342", "I63343", "I63349", "I634", "I6340", "I6341", "I63411", "I63412", "I63413", "I63419", "I6342", "I63421", "I63422", "I63423", "I63429", "I6343", "I63431", "I63432", "I63433", "I63439", "I6344", "I63441", "I63442", "I63443", "I63449", "I6349", "I635", "I6350", "I6351", "I63511", "I63512", "I63513", "I63519", "I6352", "I63521", "I63522", "I63523", "I63529", "I6353", "I63531", "I63532", "I63533", "I63539", "I6354", "I63541", "I63542", "I63543", "I63549", "I6359", "I636", "I638", "I6381", "I6389", "I639", "I65", "I650", "I6501", "I6502", "I6503", "I6509", "I651", "I652", "I6521", "I6522", "I6523", "I6529", "I658", "I659", "I66", "I660", "I6601", "I6602", "I6603", "I6609", "I661", "I6611", "I6612", "I6613", "I6619", "I662", "I6621", "I6622", "I6623", "I6629", "I663", "I668", "I669"

### All relevant diagnostic codes with blood clot
"43400", "43401", "43410", "43411", "I630", "I6300", "I63011", "I63012", "I63013", "I63019", "I6302", "I6303", "I63031", "I63032", "I63033", "I63039", "I6309", "I631", "I6310", "I6311", "I63111", "I63112", "I63113", "I63119", "I6312", "I6313", "I63131", "I63132", "I63133", "I63139", "I6319", "I633", "I6330", "I6331", "I63311", "I63312", "I63313", "I63319", "I6332", "I63321", "I63322", "I63323", "I63329", "I6333", "I63331", "I63332", "I63333", "I63339", "I6334", "I63341", "I63342", "I63343", "I63349", "I6339", "I634", "I6340", "I6341", "I63411", "I63412", "I63413", "I63419", "I6342", "I63421", "I63422", "I63423", "I63429", "I6343", "I63431", "I63432", "I63433", "I63439", "I6344", "I63441", "I63442", "I63443", "I63449", "I6349"

## Required features and where to find them
Age, gender, ED entrance time, ED departure time, discharge time, admission time, ct scan/portable ct scan time, TPA time, icu in time, icu out time, neurologist time, neuro icu, SLP, sworker, physio time

- hadm_id = hospital admission id => one patient can be admitted multiple times
- TPA = Alteplase drug
____________________________________________________________________________________________________
MISSING FEATURES ATM:
- Dietitian
- Occupational therapist
- Physio
____________________________________________________________________________________________________
- Age                           mimiciv_hosp/patients
- gender:                       mimiciv_hosp/patients
- ED entrance time:             mimiciv_hosp/transfers/intime
- ED leave time:                mimiciv_hosp/transfers/outtime
- admission time:               mimiciv_hosp/admissions
- discharge time:               mimiciv_hosp/admissions
- CT Scan time:                 itemid 221214 in mimiciv_icu/procedureevents
- portable ct scan:             itemid 229582 in mimiciv_icu/procedureevents
- TPA time:                     refer to the queries below for this
- ICU time in:                  mimiciv_icu/icustays
- ICU out time:                 mimiciv_icu/icustays
- Neurologist:                  mimiciv_hosp/transfers
- Speech language pathologist:  itemid 229298 in mimiciv_icu/procedureevents (also known as "speech and swallow evaluation")
- SWorker:                      itemid 225078 mimiciv_icu/chartevents
- Physio/physical therapist:    itemid 223871 in mimiciv_icu/chartevents
- item id:                      mimiciv_icu/d_items

# TPA sourcing
tPA drugs: 
- inputevents.csv: subject_id, hadm_id, starttime, endtime, itemid = 221319
    - 94 records for alteplase 
- prescriptions.csv: drug, subject_id, hadm_id, starttime, stoptime
    - drug: 
        - Alteplase
        - Alteplase (Catherer Clearance)
        - Alteplase 1mg/1mL ( Clearance ie. Temp TLCL )
        - Alteplase 1mg/Flush Volume ( Dialysis/Pheresis Catheters )
        - Alteplase 1mg/5mL ( Clearance ie. Indwelling port )
    - 1451 records for alteplase
- pharmacy.csv: medication, subject_id, hadm_id, starttime, stoptime
    - medication: 
        - Alteplase
        - Alteplase (Catheter Clearance)
        - Alteplase 1mg/1mL ( Clearance ie. Temp TLCL )
        - Alteplase 1mg/Flush Volume ( Dialysis/Pheresis Catheters )
        - Alteplase 1mg/2mL ( Clearance ie. PICC, midline, tunneled access line, PA )
        - Alteplase 1mg/2mL ( Clearance ie. PICC, tunneled access line )
        - Alteplase 1mg/2mL ( Clearance ie. PICC, tunneled access line, PA )
        - Alteplase 1mg/5mL ( Clearance ie. Indwelling port )
    - 1572 records for alteplase
- emar.csv: medication, subject_id, hadm_id, scheduletime, storetime
    - medication: same as pharmacy.csv
    - 452 records for alteplase
- drgcodes.csv: subject_id, hadm_id, drg_code, has no time administered XXX
    - drg_code:
        - 61
        - 62
        - 63

# Step 2: Gathering and combining data from different tables

## Bigquery
### table extractions
CREATE TABLE `capstone-423022.TableExtracts.selected_conditions` AS 
SELECT DISTINCT subject_id, hadm_id, icd_code 
FROM `physionet-data.mimiciv_hosp.diagnoses_icd` 
WHERE icd_code IN ("43300", "43301", "43310", "43311", "43320", "43321", "43330", "43331", "43380", "43381", "43390", "43391", "43400", "43401", "43410", "43411", "43490", "43491", "436", "4370", "4371", "4376", "I63", "I6300", "I6301", "I63011", "I63012", "I63013", "I63019", "I6302", "I6303", "I63031", "I63032", "I63033", "I63039", "I6309", "I631", "I6310", "I6311", "I63111", "I63112", "I63113", "I63119", "I6312", "I6313", "I63131", "I63132", "I63133", "I63139", "I6319", "I632", "I6320", "I6321", "I63211", "I63212", "I63213", "I63219", "I6322", "I6323", "I63231", "I63232", "I63233", "I63239", "I6329", "I633", "I6330", "I6331", "I63311", "I63312", "I63313", "I63319", "I6332", "I63321", "I63322", "I63323", "I63329", "I6333", "I63331", "I63332", "I63333", "I63339", "I6334", "I63341", "I63342", "I63343", "I63349", "I634", "I6340", "I6341", "I63411", "I63412", "I63413", "I63419", "I6342", "I63421", "I63422", "I63423", "I63429", "I6343", "I63431", "I63432", "I63433", "I63439", "I6344", "I63441", "I63442", "I63443", "I63449", "I6349", "I635", "I6350", "I6351", "I63511", "I63512", "I63513", "I63519", "I6352", "I63521", "I63522", "I63523", "I63529", "I6353", "I63531", "I63532", "I63533", "I63539", "I6354", "I63541", "I63542", "I63543", "I63549", "I6359", "I636", "I638", "I6381", "I6389", "I639", "I65", "I650", "I6501", "I6502", "I6503", "I6509", "I651", "I652", "I6521", "I6522", "I6523", "I6529", "I658", "I659", "I66", "I660", "I6601", "I6602", "I6603", "I6609", "I661", "I6611", "I6612", "I6613", "I6619", "I662", "I6621", "I6622", "I6623", "I6629", "I663", "I668", "I669");


CREATE TABLE `capstone-423022.TableExtracts.patient_info` AS 
SELECT subject_id, gender, anchor_age 
from `physionet-data.mimiciv_hosp.patients`;


-- CREATE TABLE `capstone-423022.TableExtracts.ED_inout_time` AS 
-- SELECT subject_id, hadm_id, intime as ED_intime, outtime as ED_outtime 
-- FROM `physionet-data.mimiciv_hosp.transfers` 
-- WHERE careunit = "Emergency Department";


CREATE TABLE `capstone-423022.TableExtracts.admit_discharge` AS 
SELECT subject_id, hadm_id, admittime as admission_time, dischtime as discharge_time, edregtime as ED_intime, edouttime as ED_outtime
FROM `physionet-data.mimiciv_hosp.admissions`;


CREATE TABLE `capstone-423022.TableExtracts.CT_inout_time` AS 
SELECT subject_id, hadm_id, stay_id, starttime as CT_time
FROM `physionet-data.mimiciv_icu.procedureevents` 
WHERE itemid IN (221214, 229582); 


CREATE TABLE `capstone-423022.TableExtracts.TPA_time0` AS 
SELECT subject_id, hadm_id, starttime as TPA_time
FROM `physionet-data.mimiciv_icu.inputevents` 
WHERE itemid = 221319;


CREATE TABLE `capstone-423022.TableExtracts.TPA_time1` AS 
SELECT subject_id, hadm_id, starttime as TPA_time
FROM `physionet-data.mimiciv_hosp.prescriptions` 
WHERE drug in ("Alteplase", "Alteplase (Catherer Clearance)", "Alteplase 1mg/1mL ( Clearance ie. Temp TLCL )", "Alteplase 1mg/Flush Volume ( Dialysis/Pheresis Catheters )", "Alteplase 1mg/5mL ( Clearance ie. Indwelling port )");


CREATE TABLE `capstone-423022.TableExtracts.TPA_time2` AS 
SELECT subject_id, hadm_id, starttime as TPA_time
FROM `physionet-data.mimiciv_hosp.pharmacy` 
WHERE medication in ("Alteplase", "Alteplase (Catherer Clearance)", "Alteplase 1mg/1mL ( Clearance ie. Temp TLCL )", "Alteplase 1mg/Flush Volume ( Dialysis/Pheresis Catheters )", "Alteplase 1mg/2mL ( Clearance ie. PICC, midline, tunneled access line, PA )", "Alteplase 1mg/2mL ( Clearance ie. PICC, tunneled access line )", "Alteplase 1mg/2mL ( Clearance ie. PICC, tunneled access line, PA )", "Alteplase 1mg/5mL ( Clearance ie. Indwelling port )");


CREATE TABLE `capstone-423022.TableExtracts.TPA_time` AS 
SELECT * 
FROM `capstone-423022.TableExtracts.TPA_time0`
union distinct (select *
                from `capstone-423022.TableExtracts.TPA_time1`
                union distinct
                select *
                from `capstone-423022.TableExtracts.TPA_time2`);


CREATE TABLE `capstone-423022.TableExtracts.ICU_stays` AS 
SELECT subject_id, hadm_id, stay_id, intime as ICU_intime, outtime as ICU_outtime, los as ICU_los 
FROM `physionet-data.mimiciv_icu.icustays`;


CREATE TABLE `capstone-423022.TableExtracts.SLP_time` AS 
SELECT subject_id, hadm_id, stay_id, starttime as SLP_time
FROM `physionet-data.mimiciv_icu.procedureevents` 
WHERE itemid = 229298;


CREATE TABLE `capstone-423022.TableExtracts.SW_time` AS 
SELECT subject_id, hadm_id, stay_id, charttime as SW_time
FROM `physionet-data.mimiciv_icu.chartevents` 
WHERE itemid = 225078;

### table joins
CREATE TABLE `capstone-423022.TableJoins.r1` AS 
SELECT table_left.*, table_right.gender, table_right.anchor_age 
from `capstone-423022.TableExtracts.selected_conditions` table_left
cross join `capstone-423022.TableExtracts.patient_info` table_right
where table_left.subject_id = table_right.subject_id;
-- columns: subject_id, hadm_id, icd_code, gender, anchor_age

CREATE TABLE `capstone-423022.TableJoins.r2` AS 
select table_left.*, table_right.admission_time, table_right.discharge_time, table_right.ED_intime, table_right.ED_outtime 
from `capstone-423022.TableJoins.r1` table_left
left join `capstone-423022.TableExtracts.admit_discharge` table_right
    on table_left.subject_id = table_right.subject_id 
        and table_left.hadm_id = table_right.hadm_id;
-- columns: subject_id, hadm_id, icd_code, gender, anchor_age, admission_time, discharge_time, ED_intime, ED_outtime

CREATE TABLE `capstone-423022.TableJoins.r3` AS 
select table_left.*, table_right.TPA_time
from `capstone-423022.TableJoins.r2` table_left
left join `capstone-423022.TableExtracts.TPA_time` table_right
    on table_left.subject_id = table_right.subject_id 
        and table_left.hadm_id = table_right.hadm_id; 
-- columns: subject_id, hadm_id, icd_code, gender, anchor_age, admission_time, discharge_time, ED_intime, ED_outtime, TPA_time


CREATE TABLE `capstone-423022.TableJoins.r4` AS 
select table_left.*, table_right.stay_id, table_right.CT_time
from `capstone-423022.TableJoins.r3` table_left
left join `capstone-423022.TableExtracts.CT_inout_time` table_right
    on table_left.subject_id = table_right.subject_id 
        and table_left.hadm_id = table_right.hadm_id;
-- columns: subject_id, hadm_id, icd_code, gender, anchor_age, admission_time, discharge_time, ED_intime, ED_outtime, TPA_time, stay_id, CT_time

-- stay_id is added above. check for any issues.

CREATE TABLE `capstone-423022.TableJoins.r5` AS 
select table_left.*, table_right.ICU_intime, table_right.ICU_outtime, table_right.ICU_los
from `capstone-423022.TableJoins.r4` table_left
left join `capstone-423022.TableExtracts.ICU_stays` table_right
    on table_left.subject_id = table_right.subject_id 
        and table_left.hadm_id = table_right.hadm_id 
        and table_left.stay_id = table_right.stay_id;
-- columns: subject_id, hadm_id, icd_code, gender, anchor_age, admission_time, discharge_time, ED_intime, ED_outtime, TPA_time, stay_id, CT_time, ICU_intime, ICU_outtime, ICU_los

CREATE TABLE `capstone-423022.TableJoins.r6` AS 
select table_left.*, table_right.SLP_time
from `capstone-423022.TableJoins.r5` table_left
left join `capstone-423022.TableExtracts.SLP_time` table_right
    on table_left.subject_id = table_right.subject_id 
        and table_left.hadm_id = table_right.hadm_id 
        and table_left.stay_id = table_right.stay_id;
-- columns: subject_id, hadm_id, icd_code, gender, anchor_age, admission_time, discharge_time, ED_intime, ED_outtime, TPA_time, stay_id, CT_time, ICU_intime, ICU_outtime, ICU_los, SLP_time

CREATE TABLE `capstone-423022.TableJoins.r7` AS 
select table_left.*, table_right.SW_time
from `capstone-423022.TableJoins.r6` table_left
left join `capstone-423022.TableExtracts.SW_time` table_right
    on table_left.subject_id = table_right.subject_id 
        and table_left.hadm_id = table_right.hadm_id 
        and table_left.stay_id = table_right.stay_id;
-- columns: subject_id, hadm_id, icd_code, gender, anchor_age, admission_time, discharge_time, ED_intime, ED_outtime, TPA_time, stay_id, CT_time, ICU_intime, ICU_outtime, ICU_los, SLP_time, SW_time

In [2]:
import pandas as pd

# codes with blood clot
codes = ["43400", "43401", "43410", "43411", "I630", "I6300", "I63011", "I63012", "I63013", "I63019", "I6302", "I6303", "I63031", "I63032", "I63033", "I63039", "I6309", "I631", "I6310", "I6311", "I63111", "I63112", "I63113", "I63119", "I6312", "I6313", "I63131", "I63132", "I63133", "I63139", "I6319", "I633", "I6330", "I6331", "I63311", "I63312", "I63313", "I63319", "I6332", "I63321", "I63322", "I63323", "I63329", "I6333", "I63331", "I63332", "I63333", "I63339", "I6334", "I63341", "I63342", "I63343", "I63349", "I6339", "I634", "I6340", "I6341", "I63411", "I63412", "I63413", "I63419", "I6342", "I63421", "I63422", "I63423", "I63429", "I6343", "I63431", "I63432", "I63433", "I63439", "I6344", "I63441", "I63442", "I63443", "I63449", "I6349"]

In [3]:
res = pd.read_csv("queryResults0.5.csv")
res = res.drop_duplicates()
res[['admission_time', 'ED_intime', 'ED_outtime', 'discharge_time', 'TPA_time', 'CT_time', 'ICU_intime', 'ICU_outtime', 'SLP_time', 'SW_time']] = res[['admission_time', 'ED_intime', 'ED_outtime', 'discharge_time', 'TPA_time', 'CT_time', 'ICU_intime', 'ICU_outtime', 'SLP_time', 'SW_time']].astype('datetime64[ns]')
res['stay_id'] = res['stay_id'].astype(pd.Int64Dtype())
res = res[['subject_id', 'hadm_id', 'stay_id', 'icd_code', 'gender', 'anchor_age',
           'admission_time', 'discharge_time', 'ED_intime', 'ED_outtime', 'TPA_time',
           'CT_time', 'ICU_intime', 'ICU_outtime', 'ICU_los', 'SLP_time', 'SW_time']]
print(res.shape)

(22125, 17)


In [104]:
selection = 'sub'
if selection == 'all':
    # all icd codes
    data = res
elif selection == 'all+':
    # all icd codes of length 7+
    data = res[res.isna().sum(axis=1) < 3]
elif selection == 'sub':
    # blood clot icd codes
    data = res[res['icd_code'].isin(codes)]
else:
    # blood clot icd codes of length 7+
    data = res[res['icd_code'].isin(codes)]
    data = data[data.isna().sum(axis=1) < 3]

indices = (data.groupby(['subject_id', 'hadm_id', 'stay_id']).nunique()['ED_intime']>1).reset_index()
asd = indices[indices['ED_intime']]
data = data[~data['subject_id'].isin(asd['subject_id'].values)]
data

subject_id   hadm_id   stay_id icd_code gender  anchor_age  \
4        15246491  29591261      <NA>   I63412      F          47   
7        15156831  22148605      <NA>   I63412      F          52   
8        13058990  26951330      <NA>    43411      F          52   
12       18090606  25643446      <NA>    43411      F          59   
29       10642913  28626703      <NA>   I63432      F          69   
...           ...       ...       ...      ...    ...         ...   
22120    13873595  23417361  32368927    43411      M          87   
22121    13873595  23417361  32368927    43411      M          87   
22125    17354581  20080344  31984623    I6312      M          68   
22126    17354581  20080344  31984623    I6312      M          68   
22127    17354581  20080344  31984623    I6312      M          68   

           admission_time      discharge_time           ED_intime  \
4     2173-04-24 18:37:00 2173-04-27 14:30:00 2173-04-24 15:58:00   
7     2140-05-28 23:47:00 2140-06-04 17:00:00 2140-05-28 19:26:00   
8     2152-09-16 23:04:00 2152-09-22 15:25:00 2152-09-16 19:47:00   
12    2132-11-11 20:18:00 2132-11-12 20:15:00 2132-11-11 10:40:00   
29    2158-05-13 19:22:00 2158-05-14 17:30:00 2158-05-13 13:25:00   
...                   ...                 ...                 ...   
22120 2172-11-05 17:24:00 2172-11-26 16:51:00 2172-11-05 14:07:00   
22121 2172-11-05 17:24:00 2172-11-26 16:51:00 2172-11-05 14:07:00   
22125 2188-08-20 01:58:00 2188-09-15 12:40:00 2188-08-19 19:51:00   
22126 2188-08-20 01:58:00 2188-09-15 12:40:00 2188-08-19 19:51:00   
22127 2188-08-20 01:58:00 2188-09-15 12:40:00 2188-08-19 19:51:00   

               ED_outtime            TPA_time             CT_time  \
4     2173-04-24 18:47:00                 NaT                 NaT   
7     2140-05-29 04:16:00                 NaT                 NaT   
8     2152-09-17 01:28:00                 NaT                 NaT   
12    2132-11-11 22:07:00                 NaT                 NaT   
29    2158-05-13 21:13:00                 NaT                 NaT   
...                   ...                 ...                 ...   
22120 2172-11-05 18:00:00                 NaT 2172-11-07 03:00:00   
22121 2172-11-05 18:00:00                 NaT 2172-11-08 23:28:00   
22125 2188-08-19 21:40:00 2188-09-13 13:00:00 2188-08-28 09:45:00   
22126 2188-08-19 21:40:00 2188-09-13 13:00:00 2188-08-31 15:30:00   
22127 2188-08-19 21:40:00 2188-09-13 13:00:00 2188-08-24 12:00:00   

               ICU_intime         ICU_outtime    ICU_los SLP_time SW_time  
4                     NaT                 NaT        NaN      NaT     NaT  
7                     NaT                 NaT        NaN      NaT     NaT  
8                     NaT                 NaT        NaN      NaT     NaT  
12                    NaT                 NaT        NaN      NaT     NaT  
29                    NaT                 NaT        NaN      NaT     NaT  
...                   ...                 ...        ...      ...     ...  
22120 2172-11-05 18:50:31 2172-11-12 17:36:07   6.948333      NaT     NaT  
22121 2172-11-05 18:50:31 2172-11-12 17:36:07   6.948333      NaT     NaT  
22125 2188-08-20 01:59:31 2188-09-07 08:41:23  18.279074      NaT     NaT  
22126 2188-08-20 01:59:31 2188-09-07 08:41:23  18.279074      NaT     NaT  
22127 2188-08-20 01:59:31 2188-09-07 08:41:23  18.279074      NaT     NaT  

[6261 rows x 17 columns]

In [ ]:
keys = (data[['subject_id', 'hadm_id']].groupby(['subject_id', 'hadm_id']).size()).reset_index()[['subject_id', 'hadm_id']]
sequences = []

singular_columns = ['subject_id', 'hadm_id', 'stay_id', 'icd_code', 'gender', 'anchor_age', 'admission_time',
                    'discharge_time', 'ED_intime', 'ED_outtime']
duplicatables = ['TPA_time', 'CT_time', 'SLP_time', 'SW_time']
stay_id_dependent_columns = ['ICU_intime', 'ICU_outtime', 'ICU_los']


for rownum in range(len(keys)):
    sequence = {}
    sub = data[(data['subject_id']==keys.iloc[rownum, 0]) & (data['hadm_id']==keys.iloc[rownum, 1])]
    for col in singular_columns:
        if not sub[col].isna().any():
            sequence[col] = sub[col].unique()[0]
            

    for col in duplicatables:
        i = 0
        uniques = sub[col].dropna().unique()
        if len(uniques) > 1:
            for var in uniques:
                sequence[f'{col}{i}'] = sub[sub[col]==var][col].values[0]
                i += 1
            
    i = 0
    for stayid in sub['stay_id'].dropna().unique():
        if len((sub[sub['stay_id']==stayid])[['ICU_intime', 'ICU_outtime', 'ICU_los']]) != 0:
            sequence[f'icuin{i}'] = sub[sub['stay_id']==stayid]['ICU_intime'].drop_duplicates().values[0]
            sequence[f'icuout{i}'] = sub[sub['stay_id']==stayid]['ICU_outtime'].drop_duplicates().values[0]
            sequence[f'iculos{i}'] = sub[sub['stay_id']==stayid]['ICU_los'].drop_duplicates().values[0]
        i += 1
    sequences.append(sequence)

In [106]:
uniq = (data.groupby(['subject_id', 'hadm_id']).nunique()).reset_index()
avg_length = uniq[['admission_time', 'discharge_time', 'ED_intime', 'ED_outtime', 'TPA_time', 'CT_time', 'SLP_time', 'SW_time']].sum(axis=1).mean()
print(len(sequences))
print(avg_length)

3267
4.624119987756352


# Sequence metadata
- A patient who is admitted to the ICU is issued a stay_id 
- for each stay_id, social workers can visit many times (like 15 even)
- for some combinations of subject_id and hadm_id, there exist duplicate events, such as more than 1 ct scan, icu entry, and so on.
- for every combination of subject_id and hadm_id we have 1 sequence of events.
    - for every sequence of events, we can have multiple sub-sequences of events based on stay_id which consists of ct_time, icu_intime, icu_outtime, icu_los
- icu values in the sequences are labeled as icu0, icu1, ...
- all type of icu are considered
- 

- sequence lengths for different condition codes
    - all condition codes:                                  12332 sequences of average length 3.98
    - all condition codes of length 7+:                     683   sequences of average length 8.80
    - subset of condition codes:                            3267  sequences of average length 4.62
    - subbset of condition codes with length 7+:            337   sequences of average length 8.40